File: apps/mobile/lib/features/logistics/domain/shipment.dart
import 'package:flutter/foundation.dart';
import '../../cart_checkout/domain/order.dart';

/// Granular shipment statuses for rural logistics and line-haul tracking
enum ShipmentStatus {
  created,
  pickupScheduled,
  pickedUp,
  processing,
  inTransit,
  atRegionalHub,
  outForDelivery,
  delivered,
  deliveryAttempted,
  cancelled,
  returned;

  String get displayName => switch (this) {
        ShipmentStatus.created => 'Shipment Created',
        ShipmentStatus.pickupScheduled => 'Pickup Scheduled',
        ShipmentStatus.pickedUp => 'Package Picked Up',
        ShipmentStatus.processing => 'Hub Processing',
        ShipmentStatus.inTransit => 'In Transit',
        ShipmentStatus.atRegionalHub => 'At Regional Sorting Hub',
        ShipmentStatus.outForDelivery => 'Out for Delivery',
        ShipmentStatus.delivered => 'Consignment Delivered',
        ShipmentStatus.deliveryAttempted => 'Delivery Attempted',
        ShipmentStatus.cancelled => 'Shipment Cancelled',
        ShipmentStatus.returned => 'Returned to Origin',
      };

  bool get isActive =>
      this != ShipmentStatus.delivered &&
      this != ShipmentStatus.cancelled &&
      this != ShipmentStatus.returned;

  /// Maps granular ShipmentStatus to backward-compatible OrderStatus
  OrderStatus toOrderStatus() {
    switch (this) {
      case ShipmentStatus.created:
      case ShipmentStatus.pickupScheduled:
        return OrderStatus.confirmed;
      case ShipmentStatus.pickedUp:
      case ShipmentStatus.processing:
        return OrderStatus.processing;
      case ShipmentStatus.inTransit:
      case ShipmentStatus.atRegionalHub:
        return OrderStatus.shipped;
      case ShipmentStatus.outForDelivery:
      case ShipmentStatus.deliveryAttempted:
        return OrderStatus.outForDelivery;
      case ShipmentStatus.delivered:
        return OrderStatus.delivered;
      case ShipmentStatus.cancelled:
      case ShipmentStatus.returned:
        return OrderStatus.cancelled;
    }
  }
}

@immutable
class DeliveryAgent {
  const DeliveryAgent({
    required this.id,
    required this.name,
    required this.phone,
    required this.carrier,
    required this.vehicleType,
    required this.vehicleNumber,
    required this.rating,
  });

  final String id;
  final String name;
  final String phone;
  final String carrier;
  final String vehicleType;
  final String vehicleNumber;
  final double rating;

  factory DeliveryAgent.fromJson(Map<String, dynamic> json) {
    return DeliveryAgent(
      id: json['id'] as String,
      name: json['name'] as String,
      phone: json['phone'] as String,
      carrier: json['carrier'] as String? ?? 'Delhivery Rural Express',
      vehicleType: json['vehicle_type'] as String? ?? 'EV Three-Wheeler',
      vehicleNumber: json['vehicle_number'] as String? ?? 'MH-12-TR-4921',
      rating: (json['rating'] as num?)?.toDouble() ?? 4.8,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'phone': phone,
        'carrier': carrier,
        'vehicle_type': vehicleType,
        'vehicle_number': vehicleNumber,
        'rating': rating,
      };
}

@immutable
class TrackingEvent {
  const TrackingEvent({
    required this.id,
    required this.shipmentId,
    required this.status,
    required this.location,
    required this.description,
    required this.eventTime,
  });

  final String id;
  final String shipmentId;
  final ShipmentStatus status;
  final String location;
  final String description;
  final DateTime eventTime;

  factory TrackingEvent.fromJson(Map<String, dynamic> json) {
    return TrackingEvent(
      id: json['id'] as String,
      shipmentId: json['shipment_id'] as String,
      status: ShipmentStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => ShipmentStatus.created,
      ),
      location: json['location'] as String,
      description: json['description'] as String,
      eventTime: DateTime.parse(json['event_time'] as String),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'shipment_id': shipmentId,
        'status': status.name,
        'location': location,
        'description': description,
        'event_time': eventTime.toIso8601String(),
      };
}

@immutable
class DeliveryAttempt {
  const DeliveryAttempt({
    required this.id,
    required this.shipmentId,
    required this.attemptNumber,
    required this.status,
    required this.reason,
    required this.attemptedAt,
    this.notes,
    this.nextAttemptDate,
  });

  final String id;
  final String shipmentId;
  final int attemptNumber;
  final String status;
  final String reason;
  final DateTime attemptedAt;
  final String? notes;
  final DateTime? nextAttemptDate;

  factory DeliveryAttempt.fromJson(Map<String, dynamic> json) {
    return DeliveryAttempt(
      id: json['id'] as String,
      shipmentId: json['shipment_id'] as String,
      attemptNumber: json['attempt_number'] as int? ?? 1,
      status: json['status'] as String? ?? 'rescheduled',
      reason: json['reason'] as String,
      attemptedAt: DateTime.parse(json['attempted_at'] as String),
      notes: json['notes'] as String?,
      nextAttemptDate: json['next_attempt_date'] != null
          ? DateTime.parse(json['next_attempt_date'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'shipment_id': shipmentId,
        'attempt_number': attemptNumber,
        'status': status,
        'reason': reason,
        'attempted_at': attemptedAt.toIso8601String(),
        'notes': notes,
        'next_attempt_date': nextAttemptDate?.toIso8601String(),
      };
}

@immutable
class Shipment {
  const Shipment({
    required this.id,
    required this.orderId,
    required this.userId,
    required this.provider,
    required this.trackingNumber,
    required this.status,
    required this.originLocation,
    required this.destinationLocation,
    required this.currentLocation,
    required this.estimatedDeliveryStart,
    required this.estimatedDeliveryEnd,
    required this.serviceZone,
    required this.distanceBand,
    required this.createdAt,
    required this.updatedAt,
    this.deliveryAgent,
    this.attempts = const [],
    this.events = const [],
  });

  final String id;
  final String orderId;
  final String userId;
  final String provider;
  final String trackingNumber;
  final ShipmentStatus status;
  final String originLocation;
  final String destinationLocation;
  final String currentLocation;
  final DateTime estimatedDeliveryStart;
  final DateTime estimatedDeliveryEnd;
  final String serviceZone;
  final String distanceBand;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DeliveryAgent? deliveryAgent;
  final List<DeliveryAttempt> attempts;
  final List<TrackingEvent> events;

  Shipment copyWith({
    String? id,
    String? orderId,
    String? userId,
    String? provider,
    String? trackingNumber,
    ShipmentStatus? status,
    String? originLocation,
    String? destinationLocation,
    String? currentLocation,
    DateTime? estimatedDeliveryStart,
    DateTime? estimatedDeliveryEnd,
    String? serviceZone,
    String? distanceBand,
    DateTime? createdAt,
    DateTime? updatedAt,
    DeliveryAgent? deliveryAgent,
    List<DeliveryAttempt>? attempts,
    List<TrackingEvent>? events,
  }) {
    return Shipment(
      id: id ?? this.id,
      orderId: orderId ?? this.orderId,
      userId: userId ?? this.userId,
      provider: provider ?? this.provider,
      trackingNumber: trackingNumber ?? this.trackingNumber,
      status: status ?? this.status,
      originLocation: originLocation ?? this.originLocation,
      destinationLocation: destinationLocation ?? this.destinationLocation,
      currentLocation: currentLocation ?? this.currentLocation,
      estimatedDeliveryStart: estimatedDeliveryStart ?? this.estimatedDeliveryStart,
      estimatedDeliveryEnd: estimatedDeliveryEnd ?? this.estimatedDeliveryEnd,
      serviceZone: serviceZone ?? this.serviceZone,
      distanceBand: distanceBand ?? this.distanceBand,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      deliveryAgent: deliveryAgent ?? this.deliveryAgent,
      attempts: attempts ?? this.attempts,
      events: events ?? this.events,
    );
  }
}

File: apps/web/features/logistics/domain/shipment.ts
// ============================================================
// LOGISTICS & SHIPMENT DOMAIN ENTITIES
// Platform-neutral contracts shared across Web and Mobile
// ============================================================

import type { OrderStatus } from '@/types';

export type ShipmentStatus =
  | 'created'
  | 'pickupScheduled'
  | 'pickedUp'
  | 'processing'
  | 'inTransit'
  | 'atRegionalHub'
  | 'outForDelivery'
  | 'delivered'
  | 'deliveryAttempted'
  | 'cancelled'
  | 'returned';

export interface DeliveryAgent {
  id: string;
  name: string;
  phone: string;
  carrier: string;
  vehicleType: string;
  vehicleNumber: string;
  rating: number;
}

export interface TrackingEvent {
  id: string;
  shipmentId: string;
  status: ShipmentStatus;
  location: string;
  description: string;
  eventTime: string;
}

export interface DeliveryAttempt {
  id: string;
  shipmentId: string;
  attemptNumber: number;
  status: 'failed' | 'rescheduled' | 'delivered';
  reason:
    | 'customer_unavailable'
    | 'address_clarification_required'
    | 'weather_delay'
    | 'route_delay'
    | 'security_gate_locked';
  notes?: string;
  attemptedAt: string;
  nextAttemptDate?: string;
}

export interface EstimatedDeliveryWindow {
  start: string;
  end: string;
  serviceZone: string;
  ruralConfidenceScore: number; // e.g. 96 (%)
  advisory?: string;
}

export interface Shipment {
  id: string;
  orderId: string;
  userId: string;
  provider: string; // 'demo_logistics' | 'delhivery_rural' | 'shiprocket'
  trackingNumber: string;
  status: ShipmentStatus;
  originLocation: string;
  destinationLocation: string;
  currentLocation: string;
  estimatedDeliveryStart: string;
  estimatedDeliveryEnd: string;
  deliveryAgent?: DeliveryAgent;
  serviceZone: string; // e.g. 'Rural Priority Route', 'Intra-District Tier 2'
  distanceBand: string; // e.g. '145 km', 'Zone B (100–250 km)'
  attempts: DeliveryAttempt[];
  events: TrackingEvent[];
  metadata?: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export const SHIPMENT_STATUS_LABELS: Record<ShipmentStatus, string> = {
  created: 'Shipment Created',
  pickupScheduled: 'Pickup Scheduled',
  pickedUp: 'Package Picked Up',
  processing: 'Hub Processing',
  inTransit: 'In Transit',
  atRegionalHub: 'At Regional Sorting Hub',
  outForDelivery: 'Out for Delivery',
  delivered: 'Consignment Delivered',
  deliveryAttempted: 'Delivery Attempted',
  cancelled: 'Shipment Cancelled',
  returned: 'Returned to Origin',
};

export const SHIPMENT_STATUS_COLORS: Record<ShipmentStatus, { bg: string; color: string }> = {
  created: { bg: 'var(--color-info-light)', color: 'var(--color-info)' },
  pickupScheduled: { bg: 'var(--color-info-light)', color: 'var(--color-info)' },
  pickedUp: { bg: 'var(--color-info-light)', color: 'var(--color-info)' },
  processing: { bg: 'var(--color-warning-light)', color: 'var(--color-warning)' },
  inTransit: { bg: 'var(--color-amber-50)', color: 'var(--color-amber-600)' },
  atRegionalHub: { bg: 'var(--color-amber-50)', color: 'var(--color-amber-600)' },
  outForDelivery: { bg: 'var(--color-amber-50)', color: 'var(--color-amber-600)' },
  delivered: { bg: 'var(--color-success-light)', color: 'var(--color-success)' },
  deliveryAttempted: { bg: 'var(--color-warning-light)', color: 'var(--color-warning)' },
  cancelled: { bg: 'var(--color-error-light)', color: 'var(--color-error)' },
  returned: { bg: 'var(--color-error-light)', color: 'var(--color-error)' },
};

/**
 * Sequential linear progression of canonical logistics milestones for UI rendering
 */
export const SHIPMENT_PROGRESSION_STEPS: ShipmentStatus[] = [
  'created',
  'pickedUp',
  'processing',
  'inTransit',
  'atRegionalHub',
  'outForDelivery',
  'delivered',
];

/**
 * Maps a granular ShipmentStatus into the high-level backward-compatible OrderStatus
 */
export function shipmentStatusToOrderStatus(status: ShipmentStatus): OrderStatus {
  switch (status) {
    case 'created':
    case 'pickupScheduled':
      return 'confirmed';
    case 'pickedUp':
    case 'processing':
      return 'processing';
    case 'inTransit':
    case 'atRegionalHub':
      return 'shipped';
    case 'outForDelivery':
    case 'deliveryAttempted':
      return 'outForDelivery';
    case 'delivered':
      return 'delivered';
    case 'cancelled':
    case 'returned':
      return 'cancelled';
    default:
      return 'confirmed';
  }
}
